// @frogpond/ccci-carleton-college/v1/news/index.mjs
import {ONE_HOUR} from '@frogpond/ccc-lib'
import {fetchRssFeed} from '@frogpond/ccc-rss-feed'
import {fetchWpJson} from '@frogpond/ccc-wpjson-feed'
import {noonNewsBulletein} from './nnb'
import mem from 'mem'

export const cachedRssFeed = mem(fetchRssFeed, {maxAge: ONE_HOUR})
export const cachedWpJsonFeed = mem(fetchWpJson, {maxAge: ONE_HOUR})
export const cachedNoonNewsBulletein = mem(noonNewsBulletein, {
	maxAge: ONE_HOUR * 6,
})

export async function nnb(ctx) {
	ctx.body = await cachedNoonNewsBulletein()
}

export async function rss(ctx) {
	ctx.body = await cachedRssFeed(ctx.query.url)
}

export async function wpJson(ctx) {
	ctx.body = await cachedWpJsonFeed(ctx.query.url)
}

export async function carletonNow(ctx) {
	let url = 'https://apps.carleton.edu/media_relations/feeds/blogs/news'
	ctx.body = await cachedRssFeed(url)
}
export async function carletonian(ctx) {
	let url = 'https://apps.carleton.edu/carletonian/feeds/blogs/tonian'
	ctx.body = await cachedRssFeed(url)
}
export async function krlxNews(ctx) {
	let url = 'http://www.krlx.org/wp-json/wp/v2/posts/'
	// eslint-disable-next-line camelcase
	ctx.body = await cachedWpJsonFeed(url, {per_page: 10, _embed: true})
}
