// @frogpond/ccci-carleton-college/v1/news/index.mjs
import {ONE_HOUR} from '@frogpond/ccc-lib'
import {fetchRssFeed, fetchWpJson} from '@frogpond/ccc-lib/feeds'
import {noonNewsBulletein} from './nnb'
import mem from 'mem'

const GET_RSS_FEED = mem(fetchRssFeed, {maxAge: ONE_HOUR})
const GET_WP_FEED = mem(fetchWpJson, {maxAge: ONE_HOUR})

const NOON_NEWS_BULLETEIN = mem(noonNewsBulletein, {maxAge: ONE_HOUR * 6})

export async function nnb(ctx) {
	ctx.body = await NOON_NEWS_BULLETEIN()
}

export async function rss(ctx) {
	ctx.body = await GET_RSS_FEED(ctx.query.url)
}

export async function wpJson(ctx) {
	ctx.body = await GET_WP_FEED(ctx.query.url)
}

export async function carletonNow(ctx) {
	let url = 'https://apps.carleton.edu/media_relations/feeds/blogs/news'
	ctx.body = await GET_RSS_FEED(url)
}
export async function carletonian(ctx) {
	let url = 'https://apps.carleton.edu/carletonian/feeds/blogs/tonian'
	ctx.body = await GET_RSS_FEED(url)
}
export async function krlxNews(ctx) {
	let url = 'https://www.krlx.org/wp-json/wp/v2/posts/'
	// eslint-disable-next-line camelcase
	ctx.body = await GET_WP_FEED(url, {per_page: 10, _embed: true})
}
