// @frogpond/ccci-stolaf-college/v1/jobs/index.js
import {get, ONE_DAY} from '@frogpond/ccc-lib'
import mem from 'mem'
import _jsdom from 'jsdom'
import getUrls from 'get-urls'

const {JSDOM} = _jsdom

export function cleanJob(job) {
	// these all need to retain their newlines
	const description = cleanTextBlock(
		JSDOM.fragment(job.description).textContent,
	)
	const comments = cleanTextBlock(JSDOM.fragment(job.comments).textContent)
	const skills = cleanTextBlock(JSDOM.fragment(job.skills).textContent)
	const howToApply = cleanTextBlock(JSDOM.fragment(job.howToApply).textContent)
	const timeline = cleanTextBlock(JSDOM.fragment(job.timeline).textContent)
	const timeOfHours = cleanTextBlock(
		JSDOM.fragment(job.timeOfHours).textContent,
	)

	const contactEmail = fixupEmailFormat(job.contactEmail)
	const contactPhone = fixupPhoneFormat(job.contactPhone)

	const contactFirstName = job.contactFirstName
	const contactLastName = job.contactLastName
	const contactName = `${contactFirstName} ${contactLastName}`.trim()

	const links = getLinksFromJob({description, comments, skills, howToApply})

	return {
		comments: comments,
		contactEmail: contactEmail,
		contactName: contactName,
		contactPhone: contactPhone,
		description: description,
		goodForIncomingStudents: job.goodForIncomingStudents,
		hoursPerWeek: job.hoursPerWeek,
		howToApply: howToApply,
		id: job.id,
		lastModified: job.lastModified,
		links: links,
		office: job.office,
		openPositions: job.openPositions,
		skills: skills,
		timeline: timeline,
		timeOfHours: timeOfHours,
		title: job.title,
		type: job.type,
		year: job.year,
	}
}

function cleanTextBlock(text) {
	return text.replace(/\s+/g, ' ')
}

export function getLinksFromJob({description, comments, skills, howToApply}) {
	// Clean up returns, newlines, tabs, and misc symbols...
	// ...and search for application links in the text
	return Array.from(
		new Set([
			...getUrls(description),
			...getUrls(comments),
			...getUrls(skills),
			...getUrls(howToApply),
		]),
	)
}

function fixupPhoneFormat(phoneNumber) {
	return phoneNumber.length === 4 ? `507-786-${phoneNumber}` : phoneNumber
}

function fixupEmailFormat(email) {
	if (!/@/.test(email)) {
		// No @ in address ... e.g. smith
		return `${email}@stolaf.edu`
	} else if (/@$/.test(email)) {
		// @ at end ... e.g. smith@
		return `${email}stolaf.edu`
	} else {
		// Defined address ... e.g. smith@stolaf.edu
		return email
	}
}

async function _getJobs() {
	let url = 'https://www.stolaf.edu/apps/stuwork/index.cfm'
	let searchParams = {fuseaction: 'getall', nostructure: '1'}

	let data = await get(url, {searchParams}).json()

	let cleaned = data.map(cleanJob)

	return cleaned
}

export const getJobs = mem(_getJobs, {maxAge: ONE_DAY})

export async function jobs(ctx) {
	ctx.cacheControl(ONE_DAY)

	ctx.body = await getJobs()
}
