// @frogpond/ccci-carleton-college/v1/menu/index.mjs
import {get, ONE_DAY} from '@frogpond/ccc-lib'
import {cafe as getBonAppCafe, menu as getBonAppMenu} from '@frogpond/ccc-lib/bonapp'
import mem from 'mem'

const pauseMenuUrl = 'https://stodevx.github.io/AAO-React-Native/pause-menu.json'
const GET_DAY = mem(get, {maxAge: ONE_DAY})
export const getPauseMenu = () => GET_DAY(pauseMenuUrl, {json: true})

const getMenu = mem(getBonAppMenu, {maxAge: ONE_DAY})
const getInfo = mem(getBonAppCafe, {maxAge: ONE_DAY})

export const getCafe = cafeId =>
	Promise.all([getMenu(cafeId), getInfo(cafeId)])

export async function pauseMenu(ctx) {
	ctx.body = await getPauseMenu().body
}

export async function bonAppMenu(ctx) {
	ctx.body = await getBonAppMenu(ctx.params.cafeId)
}

export async function bonAppCafe(ctx) {
	ctx.body = await getBonAppCafe(ctx.params.cafeId)
}

let cafeIds = {
	stav: 261,
	cage: 262,
	kingsRoom: 263,
	burton: 35,
	ldc: 36,
	sayles: 34,
	weitz: 458,
}

export async function stavCafe(ctx) {
	ctx.body = await getBonAppCafe(cafeIds.stav)
}

export async function stavMenu(ctx) {
	ctx.body = await getBonAppMenu(cafeIds.stav)
}

export async function cageCafe(ctx) {
	ctx.body = await getBonAppCafe(cafeIds.cage)
}

export async function cageMenu(ctx) {
	ctx.body = await getBonAppMenu(cafeIds.cage)
}

export async function kingsRoomCafe(ctx) {
	ctx.body = await getBonAppCafe(cafeIds.kingsRoom)
}

export async function kingsRoomMenu(ctx) {
	ctx.body = await getBonAppMenu(cafeIds.kingsRoom)
}

export async function burtonCafe(ctx) {
	ctx.body = await getBonAppCafe(cafeIds.burton)
}

export async function burtonMenu(ctx) {
	ctx.body = await getBonAppMenu(cafeIds.burton)
}

export async function ldcCafe(ctx) {
	ctx.body = await getBonAppCafe(cafeIds.ldc)
}

export async function ldcMenu(ctx) {
	ctx.body = await getBonAppMenu(cafeIds.ldc)
}

export async function saylesCafe(ctx) {
	ctx.body = await getBonAppCafe(cafeIds.sayles)
}

export async function saylesMenu(ctx) {
	ctx.body = await getBonAppMenu(cafeIds.sayles)
}

export async function weitzCafe(ctx) {
	ctx.body = await getBonAppCafe(cafeIds.weitz)
}

export async function weitzMenu(ctx) {
	ctx.body = await getBonAppMenu(cafeIds.weitz)
}
