// @frogpond/ccci-carleton-college/v1/menu/index.js
import {get, ONE_HOUR} from '@frogpond/ccc-lib'
import * as bonapp from '@frogpond/ccc-bonapp'
import mem from 'memoize'

const pauseMenuUrl =
	'https://stodevx.github.io/AAO-React-Native/pause-menu.json'
const GET_LONG = mem(get, {maxAge: ONE_HOUR})
export const getPauseMenu = () => GET_LONG(pauseMenuUrl, {json: true})

const getMenu = mem(bonapp.menu, {maxAge: ONE_HOUR})
const getInfo = mem(bonapp.cafe, {maxAge: ONE_HOUR})
const getNutrition = mem(bonapp.nutrition, {maxAge: ONE_HOUR})

export const getCafe = (cafeId) =>
	Promise.all([getMenu(cafeId), getInfo(cafeId)])

export async function pauseMenu(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getPauseMenu().then((resp) => resp.body)
}

export async function bonAppMenu(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getMenu(ctx.params.cafeId)
}

export async function bonAppCafe(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getInfo(ctx.params.cafeId)
}

export async function bonAppNutrition(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getNutrition(ctx.params.itemId)
}

let cafeIds = {
	stav: '261',
	cage: '262',
	kingsRoom: '263',
	burton: '35',
	ldc: '36',
	sayles: '34',
	weitz: '458',
}

export async function stavCafe(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getInfo(cafeIds.stav)
}

export async function stavMenu(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getMenu(cafeIds.stav)
}

export async function cageCafe(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getInfo(cafeIds.cage)
}

export async function cageMenu(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getMenu(cafeIds.cage)
}

export async function kingsRoomCafe(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getInfo(cafeIds.kingsRoom)
}

export async function kingsRoomMenu(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getMenu(cafeIds.kingsRoom)
}

export async function burtonCafe(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getInfo(cafeIds.burton)
}

export async function burtonMenu(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getMenu(cafeIds.burton)
}

export async function ldcCafe(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getInfo(cafeIds.ldc)
}

export async function ldcMenu(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getMenu(cafeIds.ldc)
}

export async function saylesCafe(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getInfo(cafeIds.sayles)
}

export async function saylesMenu(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getMenu(cafeIds.sayles)
}

export async function weitzCafe(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getInfo(cafeIds.weitz)
}

export async function weitzMenu(ctx) {
	ctx.cacheControl(ONE_HOUR)

	ctx.body = await getMenu(cafeIds.weitz)
}
