// @frogpond/ccci-stolaf-college/v1/streams/index.mjs
import {get, ONE_DAY} from '@frogpond/ccc-lib'
import mem from 'mem'
import moment from 'moment'

const GET = mem(get, {maxAge: ONE_DAY})

export async function getStreams({streamClass, sort, dateFrom, dateTo}) {
	const params = {
		class: streamClass,
		sort: sort,
		// eslint-disable-next-line camelcase
		date_from: dateFrom,
		// eslint-disable-next-line camelcase
		date_to: dateTo,
	}

	const url = 'https://www.stolaf.edu/multimedia/api/collection'
	const data = await GET(url, {json: true, query: params}).then(
		resp => resp.body,
	)
	const processed = data.results.map(stream => {
		let {starttime} = stream
		return {
			...stream,
			starttime: moment(starttime, 'YYYY-MM-DD').toISOString(),
			date: moment(starttime, 'YYYY-MM-DD HH:mm'),
		}
	})

	return processed
}

export async function upcoming(ctx) {
	const {
		date_from: dateFrom = moment().format('YYYY-MM-DD'),
		date_to: dateTo = moment()
			.add(2, 'month')
			.format('YYYY-MM-DD'),
		sort: sortBy = 'ascending',
	} = ctx.query

	ctx.body = await getStreams({
		streamClass: 'current',
		dateFrom,
		dateTo,
		sortBy,
	})
}

export async function archived(ctx) {
	const {
		date_to: dateTo = moment().format('YYYY-MM-DD'),
		date_from: dateFrom = moment()
			.subtract(2, 'month')
			.format('YYYY-MM-DD'),
		sort: sortBy = 'ascending',
	} = ctx.query

	ctx.body = await getStreams({
		streamClass: 'archived',
		dateFrom,
		dateTo,
		sortBy,
	})
}
