// @frogpond/ccci-carleton-college/v1/convos/index.js
import {get, ONE_HOUR, makeAbsoluteUrl} from '@frogpond/ccc-lib'
import {fromHtml} from '@frogpond/ccc-markdown'
import mem from 'mem'
import _jsdom from 'jsdom'
import moment from 'moment'
const {JSDOM} = _jsdom

const archiveBase =
	'https://apps.carleton.edu/events/convocations/feeds/media_files?page_id=342645'

function processConvo(event) {
	let title = JSDOM.fragment(
		event.querySelector('title').textContent,
	).textContent.trim()

	let description = event.querySelector('description')
	description = description
		? JSDOM.fragment(
				event.querySelector('description').textContent,
			).textContent.trim()
		: ''

	let pubDate = moment(event.querySelector('pubDate').textContent)

	let enclosure = event.querySelector('enclosure')
	enclosure = enclosure
		? {
				type: enclosure.getAttribute('type')
					? enclosure.getAttribute('type')
					: '',
				url: enclosure.getAttribute('url') ? enclosure.getAttribute('url') : '',
				length: enclosure.getAttribute('length')
					? enclosure.getAttribute('length')
					: '',
			}
		: null

	return {title, description, pubDate, enclosure}
}

async function fetchUpcoming(eventId) {
	let baseUrl = 'https://apps.carleton.edu/events/convocations/'
	let url = 'https://apps.carleton.edu/events/convocations/'
	let resp = await get(url, {query: {event_id: String(eventId)}})

	let dom = new JSDOM(resp.body)

	let eventEl = dom.window.document.querySelector('.eventItemMain')

	let descEl = eventEl.querySelector('.eventContent')
	let descText = descEl ? fromHtml(descEl, {baseUrl}) : ''

	let images = [...eventEl.querySelectorAll('.media a')]
		.map((imgLink) => imgLink.getAttribute('href'))
		.map((href) => makeAbsoluteUrl(href, {baseUrl}))

	let sponsor = eventEl.querySelector('.sponsorContactInfo')
	let sponsorText = sponsor ? fromHtml(sponsor, baseUrl) : ''

	return {
		images,
		content: descText,
		sponsor: sponsorText,
	}
}

export const getUpcoming = mem(fetchUpcoming, {maxAge: ONE_HOUR * 6})

export async function upcomingDetail(ctx) {
	ctx.cacheControl(ONE_HOUR * 6)

	ctx.body = await getUpcoming(ctx.params.id)
}

async function fetchArchived() {
	let resp = await get(archiveBase)
	let dom = new JSDOM(resp.body, {contentType: 'text/xml'})
	let convos = Array.from(
		dom.window.document.querySelectorAll('rss channel item'),
	).map(processConvo)
	convos = convos.slice(0, 100)
	return Promise.all(convos)
}

export const getArchived = mem(fetchArchived, {maxAge: ONE_HOUR * 6})

export async function archived(ctx) {
	ctx.cacheControl(ONE_HOUR * 6)
	ctx.body = await getArchived()
}
