// @frogpond/ccc-lib/calendar/google.mjs
import {get} from '../http'
import moment from 'moment'

function convertGoogleEvents(data, now = moment()) {
	let events = data.map(event => {
		const startTime = moment(event.start.date || event.start.dateTime)
		const endTime = moment(event.end.date || event.end.dateTime)

		return {
			dataSource: 'google',
			startTime,
			endTime,
			title: event.summary || '',
			description: event.description || '',
			location: event.location || '',
			isOngoing: startTime.isBefore(now, 'day'),
			config: {
				startTime: true,
				endTime: true,
				subtitle: 'location',
			},
		}
	})

	return events
}

export async function googleCalendar(calendarId, now = moment()) {
	let calendarUrl = `https://www.googleapis.com/calendar/v3/calendars/${calendarId}/events`

	let params = {
		maxResults: 50,
		orderBy: 'startTime',
		showDeleted: false,
		singleEvents: true,
		timeMin: now.toISOString(),
		key: process.env.GOOGLE_CALENDAR_API_KEY,
	}

	let resp = await get(calendarUrl, {json: true, query: params})

	return convertGoogleEvents(resp.body.items)
}
