// @frogpond/ccci-stolaf-college/v1/graphql/index.mjs
import graphql from 'graphql-tools'
import * as menus from '../menu/graphql'
import * as calendars from '../calendar/graphql'
import * as majors from '../majors/graphql'
import * as departments from '../departments/graphql'
import * as dictionary from '../dictionary/graphql'
const {makeExecutableSchema} = graphql

const typeDef = `
	type Query {
		institution: String
	}

	#type Mutation {}

	schema {
		query: Query
		#mutation: Mutation
	}
`

const baseResolvers = {
	institution: () => process.env.INSTITUTION,
}

// Put together a schema
const schema = makeExecutableSchema({
	typeDefs: [
		typeDef,
		...menus.typeDefs,
		...calendars.typeDefs,
		...majors.typeDefs,
		...departments.typeDefs,
		...dictionary.typeDefs,
	],
	resolvers: Object.assign(
		{},
		calendars.resolverMap || {},
		menus.resolverMap || {},
		majors.resolverMap || {},
		departments.resolverMap || {},
		dictionary.resolverMap || {},
		{
			Query: Object.assign(
				{},
				baseResolvers,
				calendars.resolvers,
				menus.resolvers,
				majors.resolvers,
				departments.resolvers,
				dictionary.resolvers,
			),
		},
	),
})

export {schema}
