// @frogpond/ccc-google-calendar/index.js
import {get, parseHtml} from '@frogpond/ccc-lib'
import moment from 'moment'
import getUrls from 'get-urls'

function convertGoogleEvents(data, now = moment()) {
	let events = data.map((event) => {
		const title = parseHtml(event.summary || '')
		const startTime = moment(event.start.date || event.start.dateTime)
		const endTime = moment(event.end.date || event.end.dateTime)
		let description = cleanTextBlock(description || '')

		return {
			dataSource: 'google',
			startTime,
			endTime,
			title,
			description: description,
			location: event.location || '',
			isOngoing: startTime.isBefore(now, 'day'),
			links: [...getUrls(description)],
			config: {
				startTime: true,
				endTime: true,
				subtitle: 'location',
			},
		}
	})

	return events
}

function cleanTextBlock(text) {
	return parseHtml(text).replace(/\t/g, ' ').replace('<br>', '\n').trim()
}

export async function googleCalendar(calendarId, now = moment()) {
	let calendarUrl = `https://www.googleapis.com/calendar/v3/calendars/${calendarId}/events`

	let params = {
		maxResults: 50,
		orderBy: 'startTime',
		showDeleted: false,
		singleEvents: true,
		timeMin: now.toISOString(),
		key: process.env.GOOGLE_CALENDAR_API_KEY,
	}

	let resp = await get(calendarUrl, {json: true, query: params})

	return convertGoogleEvents(resp.body.items)
}
