// @frogpond/ccci-carleton-college/v1/calendar/index.mjs
import {googleCalendar, reasonCalendar} from '@frogpond/ccc-lib/calendar'
import {ONE_MINUTE} from '@frogpond/ccc-lib'
import mem from 'mem'

const GET_GOOGLE_CALENDAR = mem(googleCalendar, {maxAge: ONE_MINUTE})
const GET_REASON_CALENDAR = mem(reasonCalendar, {maxAge: ONE_MINUTE})

export async function google(ctx) {
	let {id: calendarId} = ctx.query
	ctx.body = await GET_GOOGLE_CALENDAR(calendarId)
}

export async function reason(ctx) {
	let {url: calendarUrl} = ctx.query
	ctx.body = await GET_REASON_CALENDAR(calendarUrl)
}

export async function ics(ctx) {
	ctx.throw(501, 'ICS support is not implemented yet.')
}

export async function carleton(ctx) {
	let url = 'https://apps.carleton.edu/calendar/'
	ctx.body = await GET_REASON_CALENDAR(url)
}

export async function cave(ctx) {
	let url = 'https://apps.carleton.edu/student/orgs/cave/calendar/'
	ctx.body = await GET_REASON_CALENDAR(url)
}

export async function stolaf(ctx) {
	let id = '5g91il39n0sv4c2bjdv1jrvcpq4ulm4r@import.calendar.google.com'
	ctx.body = await GET_GOOGLE_CALENDAR(id)
}

export async function northfield(ctx) {
	let id = 'thisisnorthfield@gmail.com'
	ctx.body = await GET_GOOGLE_CALENDAR(id)
}

export async function krlx(ctx) {
	let id = 'krlxradio88.1@gmail.com'
	ctx.body = await GET_GOOGLE_CALENDAR(id)
}

export async function ksto(ctx) {
	let id = 'kstonarwhal@gmail.com'
	ctx.body = await GET_GOOGLE_CALENDAR(id)
}

export async function convos(ctx) {
	let url = 'https://apps.carleton.edu/events/convocations/'
	ctx.body = await GET_REASON_CALENDAR(url)
}

export async function sumo(ctx) {
	let url = 'https://apps.carleton.edu/student/orgs/sumo/'
	ctx.body = await GET_REASON_CALENDAR(url)
}
