// @frogpond/ccci-carleton-college/v1/calendar/index.js
import {googleCalendar} from '@frogpond/ccc-google-calendar'
import {reasonCalendar} from '@frogpond/ccc-reason-calendar'
import {emsCalendar} from './ems'
import {ONE_MINUTE, ONE_HOUR} from '@frogpond/ccc-lib'
import mem from 'mem'

export const getGoogleCalendar = mem(googleCalendar, {maxAge: ONE_MINUTE})
export const getReasonCalendar = mem(reasonCalendar, {maxAge: ONE_MINUTE})
export const getEmsCalendar = mem(emsCalendar, {maxAge: ONE_HOUR})

export async function google(ctx) {
	let {id: calendarId} = ctx.query
	ctx.body = await getGoogleCalendar(calendarId)
}

export async function reason(ctx) {
	let {url: calendarUrl} = ctx.query
	ctx.body = await getReasonCalendar(calendarUrl)
}

export async function ics(ctx) {
	ctx.throw(501, 'ICS support is not implemented yet.')
}

export async function carleton(ctx) {
	let url = 'https://apps.carleton.edu/calendar/'
	ctx.body = await getReasonCalendar(url)
}

export async function ems(ctx) {
	ctx.body = await getEmsCalendar()
}

export async function cave(ctx) {
	let url = 'https://apps.carleton.edu/student/orgs/cave/calendar/'
	ctx.body = await getReasonCalendar(url)
}

export async function stolaf(ctx) {
	let id = '5g91il39n0sv4c2bjdv1jrvcpq4ulm4r@import.calendar.google.com'
	ctx.body = await getGoogleCalendar(id)
}

export async function northfield(ctx) {
	let id = 'thisisnorthfield@gmail.com'
	ctx.body = await getGoogleCalendar(id)
}

export async function krlx(ctx) {
	let id = 'krlxradio88.1@gmail.com'
	ctx.body = await getGoogleCalendar(id)
}

export async function ksto(ctx) {
	let id = 'kstonarwhal@gmail.com'
	ctx.body = await getGoogleCalendar(id)
}

export async function convos(ctx) {
	let url = 'https://apps.carleton.edu/events/convocations/'
	ctx.body = await getReasonCalendar(url)
}

export async function sumo(ctx) {
	let url = 'https://apps.carleton.edu/student/orgs/sumo/'
	ctx.body = await getReasonCalendar(url)
}
