// @frogpond/ccci-stolaf-college/v1/news/index.js
/* eslint-disable camelcase */

import {ONE_HOUR} from '@frogpond/ccc-lib'
import {fetchRssFeed} from '@frogpond/ccc-rss-feed'
import {fetchWpJson} from '@frogpond/ccc-wpjson-feed'
import mem from 'mem'

export const cachedRssFeed = mem(fetchRssFeed, {maxAge: ONE_HOUR})
export const cachedWpJsonFeed = mem(fetchWpJson, {maxAge: ONE_HOUR})

export async function rss(ctx) {
	ctx.body = await cachedRssFeed(ctx.query.url)
}

export async function wpJson(ctx) {
	ctx.body = await cachedWpJsonFeed(ctx.query.url)
}

export async function stolaf(ctx) {
	let url = 'https://wp.stolaf.edu/wp-json/wp/v2/posts'
	ctx.body = await cachedWpJsonFeed(url, {per_page: 10, _embed: true})
}

export async function oleville(ctx) {
	let url = 'https://oleville.com/wp-json/wp/v2/posts/'
	ctx.body = await cachedWpJsonFeed(url, {per_page: 10, _embed: true})
}

export async function politicole(ctx) {
	let url = 'https://oleville.com/politicole/wp-json/wp/v2/posts/'
	ctx.body = await cachedWpJsonFeed(url, {per_page: 10, _embed: true})
}

export async function mess(ctx) {
	let url = 'https://www.manitoumessenger.com/wp-json/wp/v2/posts/'
	ctx.body = await cachedWpJsonFeed(url, {per_page: 10, _embed: true})
}

export async function ksto(ctx) {
	let url = 'https://kstoradio.org/wp-json/wp/v2/posts/'
	ctx.body = await cachedWpJsonFeed(url, {per_page: 10, _embed: true})
}

export async function krlx(ctx) {
	let url = 'https://www.krlx.org/wp-json/wp/v2/posts/'
	ctx.body = await cachedWpJsonFeed(url, {per_page: 10, _embed: true})
}
