// @frogpond/ccci-stolaf-college/v1/calendar/index.js
import {googleCalendar} from '@frogpond/ccc-google-calendar'
import {reasonCalendar} from '@frogpond/ccc-reason-calendar'
import {ONE_MINUTE} from '@frogpond/ccc-lib'
import mem from 'mem'

export const getGoogleCalendar = mem(googleCalendar, {maxAge: ONE_MINUTE})
export const getReasonCalendar = mem(reasonCalendar, {maxAge: ONE_MINUTE})

export async function google(ctx) {
	let {id: calendarId} = ctx.query
	ctx.body = await getGoogleCalendar(calendarId)
}

export async function reason(ctx) {
	let {url: calendarUrl} = ctx.query
	ctx.body = await getReasonCalendar(calendarUrl)
}

export async function ics(ctx) {
	ctx.throw(501, 'ICS support is not implemented yet.')
}

export async function stolaf(ctx) {
	let id = '5g91il39n0sv4c2bjdv1jrvcpq4ulm4r@import.calendar.google.com'
	ctx.body = await getGoogleCalendar(id)
}

export async function oleville(ctx) {
	let id = 'stolaf.edu_fvulqo4larnslel75740vglvko@group.calendar.google.com'
	ctx.body = await getGoogleCalendar(id)
}

export async function northfield(ctx) {
	let id = 'thisisnorthfield@gmail.com'
	ctx.body = await getGoogleCalendar(id)
}

export async function krlx(ctx) {
	let id = 'krlxradio88.1@gmail.com'
	ctx.body = await getGoogleCalendar(id)
}

export async function ksto(ctx) {
	let id = 'kstonarwhal@gmail.com'
	ctx.body = await getGoogleCalendar(id)
}
